src/main/java/edu/kit/compiler/codegen/pattern/ConversionPattern.java

```diff
@@ -0,0 +1,84 @@
+package edu.kit.compiler.codegen.pattern;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+import edu.kit.compiler.codegen.MatcherState;
+import edu.kit.compiler.codegen.Operand;
+import edu.kit.compiler.codegen.Util;
+import edu.kit.compiler.intermediate_lang.Instruction;
+import firm.Mode;
+import firm.bindings.binding_irnode.ir_opcode;
+import firm.nodes.Node;
+import lombok.AccessLevel;
+import lombok.RequiredArgsConstructor;
+
+public class ConversionPattern implements Pattern<InstructionMatch> {
+
+    public final Pattern<OperandMatch<Operand.Register>> pattern = OperandPattern.register();
+
+    @Override
+    public InstructionMatch match(Node node, MatcherState matcher) {
+        if (node.getOpCode() == ir_opcode.iro_Conv) {
+            var pred = node.getPred(0);
+            var match = pattern.match(pred, matcher);
+            if (match.matches()) {
+                var from = pred.getMode();
+                var to = node.getMode();
+                var targetRegister = matcher.getNewRegister(Util.getSize(to));
+                return new ConversionMatch(node, match, targetRegister, from, to);
+            } else {
+                return InstructionMatch.none();
+            }
+        } else {
+            return InstructionMatch.none();
+        }
+    }
+
+    @RequiredArgsConstructor(access = AccessLevel.PRIVATE)
+    public static final class ConversionMatch extends InstructionMatch.Basic {
+
+        private final Node node;
+        private final OperandMatch<Operand.Register> match;
+        private final int targetRegister;
+        private final Mode from;
+        private final Mode to;
+
+        @Override
+        public Node getNode() {
+            return node;
+        }
+
+        @Override
+        public List<Instruction> getInstructions() {
+            var target = Operand.register(to, targetRegister);
+            return List.of(Instruction.newOp(
+                    Util.formatCmd(getCmd(), Util.getSize(to), match.getOperand(), target),
+                    List.of(match.getOperand().get()),
+                    Optional.empty(),
+                    targetRegister));
+        }
+
+        @Override
+        public Optional<Integer> getTargetRegister() {
+            return Optional.of(targetRegister);
+        }
+
+        @Override
+        public Stream<Node> getPredecessors() {
+            return match.getPredecessors();
+        }
+
+        public String getCmd() {
+            // So far only Is -> Ls and Ls -> Is is implemented
+            if (from.equals(Mode.getIs()) && to.equals(Mode.getLs())) {
+                return "movsl";
+            } else if (from.equals(Mode.getLs()) && to.equals(Mode.getIs())) {
+                return "mov";
```

Comment:
This won't work: because the size of the input register is `QUAD`, the register allocation would generate something like `movl %rax, %eax`, which is illegal. If you want to explicitely generate downward-conversions, we would either need a special case in `Instruction` or you need to use something like `lea`.

However, as far as I know this currently applies only to `div` instructions, where the register allocation already creates the appropriate move for the target register implicitly. So you should be fine by just not creating an instruction at all in this case

Reply:
You're right, this is currently generating illegal code. I think your suggestion of making this a noop and changing the size of division target registers to `DOUBLE` should work.
Note that this will generate a _useless_ temporary register due to the `Conv` node after `Div`. Would that be an issue for the register allocation?

Reply:
No, any virtual registers that aren't used will be ignored by the register allocation. This works even for the trivial allocator

Reply:
So, the original solution would not have worked with a noop. The `Conv` node now reuses the result register of the division. The fundamental issue seems to be that our virtual register are assigned a static size and can therefore not be used in different contexts.
If we want to fix/improve this, we could specify the desired size, every time the register is used. Alternatively, I could imagine adding a special instruction type for conversions.

Reply:
I would definitively prefer keeping static register sizes